### PR TITLE
Cursor/add resnet models for ios classification 9247

### PR DESCRIPTION
**Pull request by MadeWithStone**
# Description

Adds classification model (resnet and vit) support to roboflow-swift for running resnet models trained on roboflow-train on iOS.

## Type of change

Please delete options that are not relevant.

-   [ ] Bug fix (non-breaking change which fixes an issue)
-   [x] New feature (non-breaking change which adds functionality)
-   [ ] This change requires a documentation update

## How has this change been tested, please provide a testcase or example of how you tested the change?

Tested using testing suite and by running inference on images.

## Any specific deployment considerations

For example, documentation changes, usability, usage/costs, secrets, etc.

## Docs

-   [ ] Docs updated? What were the changes:

### CODE REVIEW

File: Sources/Roboflow/Classes/RFClassificationModel.swift [hunk 1]
```diff
@@ -0,0 +1,285 @@
...
+        } catch {
+            return error
+        }
+        return nil
+    }
+    
+    //Load a local model file (for manually placed models like ResNet)
+    public func loadLocalModel(modelPath: URL) -> Error? {
+        do {
+            if #available(macOS 10.14, *) {
```
[the remark this answers is not in the document]
Comment (MadeWithStone) [reply Sources/Roboflow/Classes/RFClassificationModel.swift:58]: the `*` means every version of iOS is supported

File: Sources/Roboflow/Classes/RFClassificationModel.swift [hunk 2]
```diff
@@ -0,0 +1,285 @@
...
+            
+            // Handle different output types
+            if let multiArray = output.multiArrayValue {
+                // Multi-array output (may be logits that need softmax conversion)
+                let rawValues = multiArray.dataPointer.bindMemory(to: Float.self, capacity: multiArray.count)
+                
+                // Check if values are logits (outside 0-1 range) and need softmax
+                var needsSoftmax = false
+                for i in 0..<multiArray.count {
+                    if rawValues[i] < 0.0 || rawValues[i] > 1.0 {
```
Comment (probicheaux) [inline Sources/Roboflow/Classes/RFClassificationModel.swift:146]: this definitely could return a wrong value of if it needs softmax or not -- i should think this would just be set on a per architecture basis

File: Sources/Roboflow/Classes/RFClassificationModel.swift [hunk 3]
```diff
@@ -0,0 +1,285 @@
...
+                    }
+                }
+                
+                let probabilities: [Float]
+                if needsSoftmax {
+                    // Apply softmax to convert logits to probabilities
+                    probabilities = applySoftmax(logits: rawValues, count: multiArray.count)
+                } else {
+                    // Values are already probabilities
+                    probabilities = Array(UnsafeBufferPointer(start: rawValues, count: multiArray.count))
```
Comment (probicheaux) [inline Sources/Roboflow/Classes/RFClassificationModel.swift:158]: this doesnt seem to handle the multiclass classification case (would apply sigmoid instead of softmax)

File: Sources/Roboflow/Classes/Roboflow.swift [hunk 4]
```diff
@@ -154,8 +170,20 @@ public class RoboflowMobile: NSObject {
...
-            let classes = coreMLDict["classes"] as? [String]
-            
+            var classes = coreMLDict["classes"] as? [String]
+
+            let environment = coreMLDict["environment"] as? String
+
+            // download json file at the url in `environment`
+            let environmentURL = URL(string: environment!)
+            let environmentData = try? Data(contentsOf: environmentURL!)
+            let environmentDict = try? JSONSerialization.jsonObject(with: environmentData!, options: []) as? [String: Any]
```
Comment (probicheaux) [inline Sources/Roboflow/Classes/Roboflow.swift:180]: this dict shoudl have info on if it needs softmax or sigmoid or none

File: Sources/Roboflow/Classes/RFClassificationModel.swift (hunk 3 above)
Comment (MadeWithStone) [reply Sources/Roboflow/Classes/RFClassificationModel.swift:158]: I add the sigmoid to the graph if its multiclass in which case the sum should equal 1 right?; https://github.com/roboflow/roboflow-train/blob/97f60d9b6b8bab6d2186d6476783126a206cd6a8/train/src/classification/classification_trainer.py#L318C5-L363C39

**Comment by MadeWithStone** [comment]
> Where does subtraction of means and normalization of stds happen? p sure that will need to be handled differently for resnet vs vit models

That preprocessing is done in the coreml, with the ops shown below; I just read through the ViTFeatureExtractor code and it seems to apply the same ops, though adds some other ones so I'm sure there is something to add there, but I would consider ViT out of scope (only requested classification model is resnet) so im not too worried about that, we can just specify only resnet in the docs; 

export code with normalization parameters: https://github.com/roboflow/roboflow-train/blob/97f60d9b6b8bab6d2186d6476783126a206cd6a8/train/src/classification/classification_trainer.py#L318C5-L363C39

File: Sources/Roboflow/Classes/Roboflow.swift (hunk 4 above)
Comment (MadeWithStone) [reply Sources/Roboflow/Classes/Roboflow.swift:180]: I dont see anything that jumps out as specifying whether it needs softmax, sigmoid, or none in the environment dict

File: Sources/Roboflow/Classes/RFClassificationModel.swift (hunk 3 above)
Comment (MadeWithStone) [reply Sources/Roboflow/Classes/RFClassificationModel.swift]: updated to use value passed from training, sigmoid is included in model as before

File: Sources/Roboflow/Classes/Roboflow.swift (hunk 4 above)
Comment (MadeWithStone) [reply Sources/Roboflow/Classes/Roboflow.swift:179]: multiclass only appears in a multiclass model, resolved in latest commit
